src/kernel/filesystem/fat32.zig

```diff
@@ -0,0 +1,1756 @@
+    pub fn calcCheckSum(self: *const ShortName) u8 {
+        var check_sum: u8 = 0;
+        // This is dumb, the check sum relies on the wrap around >:(
+        for (self.name) |char| {
+            check_sum = (check_sum << 7) +% (check_sum >> 1) +% char;
+        }
+        for (self.extension) |char| {
+            check_sum = (check_sum << 7) +% (check_sum >> 1) +% char;
+        }
+        return check_sum;
+    }
+};
+
+/// A complete FAT entry. This includes a list of long entries and one short entry.
+const FatDirEntry = struct {
+    /// The list of long entries. This will be in reverse order.
+    long_entry: []LongName,
+
+    /// The short entry.
+    short_entry: ShortName,
+};
+
+/// The FAT32 configuration for modern FAT32. This so to remove the redundant FAT12/16
+/// aspect of the boot sector.
+const FATConfig = struct {
+    /// The number of bytes per sector. This will be normally 512, but will depend on the
+    /// hardware.
+    bytes_per_sector: u16,
+
+    /// The number of sectors per cluster. This will depend on the size of the disk.
+    sectors_per_cluster: u8,
+
+    /// The number of reserved sectors at the beginning of the filesystem. This is normally
+    /// 32 sectors.
+    reserved_sectors: u16,
+
+    /// The number of hidden sectors. This is relevant for partitioned disks.
+    hidden_sectors: u32,
+
+    /// The total number of sectors of the filesystem.
+    total_sectors: u32,
+
+    /// The number of sectors that are occupied by the FAT.
+    sectors_per_fat: u32,
+
+    /// The cluster number of the root directory.
+    root_directory_cluster: u32,
+
+    /// The sector of the FSInfo sector.
+    fsinfo_sector: u16,
+
+    /// The sector number of the backup sector.
+    backup_boot_sector: u16,
+
+    /// If the filesystem has a FSInfo block.
+    has_fs_info: bool,
+
+    /// The number of free clusters on the disk. Will be 0xFFFFFFFF is unknown.
+    number_free_cluster: u32,
+
+    /// The next free cluster to start looking for when allocating a new cluster for a
+    /// file. Will be 0xFFFFFFFF is unknown.
+    next_free_cluster: u32,
+
+    /// The end marker when reading the cluster chain. This will be in range from
+    /// 0x0FFFFFF8 - 0x0FFFFFFF
+    cluster_end_marker: u32,
+
+    ///
+    /// Convert a cluster to the corresponding sector from the filesystems configuration.
+    ///
+    /// Arguments:
+    ///     IN self: *const FATConfig - The FAT32 configuration.
+    ///     IN cluster: u32           - The cluster number to convert.
+    ///
+    /// Return: u32
+    ///     The sector number.
+    ///
+    pub fn clusterToSector(self: *const FATConfig, cluster: u32) u32 {
+        // FAT count will be 2 as this is checked in the init function
+        return (self.sectors_per_fat * 2) + self.reserved_sectors + ((cluster - 2) * self.sectors_per_cluster);
+    }
+};
+
+///
+/// Initialise a struct from bytes.
+/// TODO: Once packed structs are good to go, then use std.mem.bytesAsValue.
```

Comment:
I've had success with packed structs in my ELF work, have you tried using them recently?

Reply:
I have did an update and when I try using packed structs, it doesn't compile and not error returned, just exit code 1. Plus the sizes are not expected. So will have to wait a bit longer.